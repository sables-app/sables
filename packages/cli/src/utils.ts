import type { SablesBuildOptions } from "./types.js";

/** @public */
export const serverTargets = {
  /**
   * Server build target for the Node.js runtime
   *
   * @public
   */
  NODE: "node",
  /**
   * Server build target for the Cloudflare Workers runtime
   *
   * @public
   */
  WORKER: "worker",
} as const;

/**
 * A utility to retrieve boolean values from the Node.js env.
 * Intended to be used within Sables configuration files.
 *
 * @public
 */
export function envFlag(name: string) {
  return !!JSON.parse(process.env[name] || "false");
}

/**
 * A utility to assist in writing configuration for Sables.
 *
 * @public
 */
export function defineConfig(config: SablesBuildOptions): SablesBuildOptions {
  return config;
}

defineConfig.serverTargets = serverTargets;
