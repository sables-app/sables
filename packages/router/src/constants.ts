/**
 * Used for SSR
 * @internal
 */
export const transitionStatuses = {
  COMPLETE: "COMPLETE",
  ERROR: "ERROR",
  INCOMPLETE: "INCOMPLETE",
  TIMEOUT: "TIMEOUT",
} as const;

/** @internal */
export const ROUTER_REDUCER_KEY = "sablesRouter";

/** @public */
export const endRouteTransitionReasons = {
  COMPLETED: "COMPLETED",
  EXITED: "EXITED",
  FAILED: "FAILED",
  INTERRUPTED: "INTERRUPTED",
} as const;

/** @internal */
export const ALL_ROUTES_KEY = "@@sablesAllRoutes";

/** @internal */
export const hookNames = {
  COMPLETE: "complete",
  END: "end",
  EXIT: "exit",
  FAILURE: "failure",
  INTERRUPT: "interrupt",
  MIDDLEWARE: "middleware",
  START: "start",
} as const;
