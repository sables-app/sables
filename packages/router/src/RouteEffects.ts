import {
  AnyPayloadAction,
  DefaultEffectAPI,
  PayloadAction,
  SYMBOL_ROUTE_EFFECTS_META,
} from "@sables/core";
import { capitalize } from "@sables/utils";

import { ALL_ROUTES_KEY, hookNames } from "./constants.js";
import { chainMiddleware, combineListeners } from "./effects.js";
import {
  EndTransitionAction,
  RouteEffectHandlers,
  RouteEffectHookName,
  RouteID,
  RouteListener,
  RouteMiddleware,
  StartTransitionAction,
} from "./types.js";

/** Actions provided to route effect handlers. */
type RouteEffectAction = StartTransitionAction | EndTransitionAction;

type RouteEffectsMeta<EffectAPI extends DefaultEffectAPI> = {
  middleware: Set<{
    referenceKey: HandlerName<RouteEffectHookName, RouteID>;
    middlewareReference: RouteMiddlewareReference<RouteID, EffectAPI, any>;
  }>;
  listeners: Set<{
    listenerReference: RouteListenerReference<RouteID, EffectAPI, any>;
  }>;
};

type HandlerName<
  H extends RouteEffectHookName,
  RID extends RouteID
> = `${Capitalize<RID>}${Capitalize<H>}`;

/** @internal */
export type HandlerMeta<
  H extends RouteEffectHookName,
  RID extends RouteID,
  Action extends AnyPayloadAction
> = [RID, H, Action];

/** @internal */
export interface RouteEffectsMethods<
  EffectAPI extends DefaultEffectAPI,
  Handlers extends HandlerMeta<
    RouteEffectHookName,
    RouteID,
    AnyPayloadAction
  > = HandlerMeta<RouteEffectHookName, never, AnyPayloadAction>
> {
  /**
   * @internal
   */
  _clone(
    mutateRouteEffectMeta?: (
      nextRouteEffectsMeta: RouteEffectsMeta<EffectAPI>
    ) => void
  ): UntouchedRouteEffects<EffectAPI, Handlers>;

  /**
   * Adds a route middleware for the given route to the end of the stack.
   *
   * @example
   *
   * const routes = createRoutes()
   *   .set("root", "/");
   *
   * createRouteEffects()
   *   .append(routes.Root.id, delayTransition(5000));
   *
   * @public
   */
  append<RID extends RouteID>(
    routeInput: RouteInput<RID>,
    effect: RouteMiddleware<StartTransitionAction, EffectAPI>
  ): UntouchedRouteEffects<
    EffectAPI,
    | HandlerMeta<typeof hookNames.MIDDLEWARE, RID, StartTransitionAction>
    | Handlers
  >;

  /**
   * Adds a route middleware for all routes to the end of the stack.
   *
   * @example
   *
   * const routes = createRoutes()
   *   .set("root", "/");
   *
   * createRouteEffects()
   *   .appendAll(delayTransition(5000));
   *
   * @public
   */
  appendAll(
    effect: RouteMiddleware<StartTransitionAction, EffectAPI>
  ): UntouchedRouteEffects<
    EffectAPI,
    | HandlerMeta<
        typeof hookNames.MIDDLEWARE,
        typeof ALL_ROUTES_KEY,
        StartTransitionAction
      >
    | Handlers
  >;

  /**
   * Adds a route middleware for the given route to the beginning of the stack.
   *
   * @example
   *
   * const routes = createRoutes()
   *   .set("root", "/");
   *
   * createRouteEffects()
   *   .prepend(routes.Root.id, delayTransition(5000));
   *
   * @public
   */
  prepend<RID extends RouteID>(
    routeInput: RouteInput<RID>,
    effect: RouteMiddleware<StartTransitionAction, EffectAPI>
  ): UntouchedRouteEffects<
    EffectAPI,
    | HandlerMeta<typeof hookNames.MIDDLEWARE, RID, StartTransitionAction>
    | Handlers
  >;

  /**
   * Adds a route middleware for all routes to the beginning of the stack.
   *
   * @example
   *
   * const routes = createRoutes()
   *   .set("root", "/");
   *
   * createRouteEffects()
   *   .prependAll(delayTransition(5000));
   *
   * @public
   */
  prependAll(
    effect: RouteMiddleware<StartTransitionAction, EffectAPI>
  ): UntouchedRouteEffects<
    EffectAPI,
    | HandlerMeta<
        typeof hookNames.MIDDLEWARE,
        typeof ALL_ROUTES_KEY,
        StartTransitionAction
      >
    | Handlers
  >;

  /**
   * Adds a route listener that's called when a route transition
   * for the given route completes.
   *
   * @example
   *
   * const routes = createRoutes()
   *   .set("root", "/");
   *
   * createRouteEffects()
   *   .onComplete(routes.Root.id, () => {
   *     console.log("The root route transition completed.");
   *   });
   *
   * @public
   */
  onComplete<RID extends RouteID>(
    routeInput: RouteInput<RID>,
    effect: RouteListener<EndTransitionAction, EffectAPI>
  ): UntouchedRouteEffects<EffectAPI, Handlers>;

  /**
   * Adds a route listener that's called when a route transition
   * for any route completes.
   *
   * @example
   *
   * const routes = createRoutes()
   *   .set("root", "/");
   *
   * createRouteEffects()
   *   .onCompleteAll(() => {
   *     console.log("A route transition completed.");
   *   });
   *
   * @public
   */
  onCompleteAll(
    effect: RouteListener<EndTransitionAction, EffectAPI>
  ): UntouchedRouteEffects<EffectAPI, Handlers>;

  /**
   * Adds a route listener that's called when a route transition
   * for the given route ends.
   *
   * @example
   *
   * const routes = createRoutes()
   *   .set("root", "/");
   *
   * createRouteEffects()
   *   .onEnd(routes.Root.id, () => {
   *     console.log("The root route transition ended.");
   *   });
   *
   * @public
   */
  onEnd<RID extends RouteID>(
    routeInput: RouteInput<RID>,
    effect: RouteListener<EndTransitionAction, EffectAPI>
  ): UntouchedRouteEffects<EffectAPI, Handlers>;

  /**
   * Adds a route listener that's called when a route transition
   * for any route ends.
   *
   * @example
   *
   * const routes = createRoutes()
   *   .set("root", "/");
   *
   * createRouteEffects()
   *   .onEndAll(() => {
   *     console.log("A route transition ended.");
   *   });
   *
   * @public
   */
  onEndAll(
    effect: RouteListener<EndTransitionAction, EffectAPI>
  ): UntouchedRouteEffects<EffectAPI, Handlers>;

  /**
   * Adds a route listener that's called when a route transition
   * for the given route exits.
   *
   * @example
   *
   * const routes = createRoutes()
   *   .set("root", "/");
   *
   * createRouteEffects()
   *   .onExit(routes.Root.id, () => {
   *     console.log("The root route transition exited.");
   *   });
   *
   * @public
   */
  onExit<RID extends RouteID>(
    routeInput: RouteInput<RID>,
    effect: RouteListener<EndTransitionAction, EffectAPI>
  ): UntouchedRouteEffects<EffectAPI, Handlers>;

  /**
   * Adds a route listener that's called when a route transition
   * for any route exits.
   *
   * @example
   *
   * const routes = createRoutes()
   *   .set("root", "/");
   *
   * createRouteEffects()
   *   .onExitAll(() => {
   *     console.log("A route transition exited.");
   *   });
   *
   * @public
   */
  onExitAll(
    effect: RouteListener<EndTransitionAction, EffectAPI>
  ): UntouchedRouteEffects<EffectAPI, Handlers>;

  /**
   * Adds a route listener that's called when a route transition
   * for the given route fails.
   *
   * @example
   *
   * const routes = createRoutes()
   *   .set("root", "/");
   *
   * createRouteEffects()
   *   .onFailure(routes.Root.id, () => {
   *     console.log("The root route transition failed.");
   *   });
   *
   * @public
   */
  onFailure<RID extends RouteID>(
    routeInput: RouteInput<RID>,
    effect: RouteListener<EndTransitionAction, EffectAPI>
  ): UntouchedRouteEffects<EffectAPI, Handlers>;

  /**
   * Adds a route listener that's called when a route transition
   * for any route fails.
   *
   * @example
   *
   * const routes = createRoutes()
   *   .set("root", "/");
   *
   * createRouteEffects()
   *   .onFailureAll(() => {
   *     console.log("A route transition failed.");
   *   });
   *
   * @public
   */
  onFailureAll(
    effect: RouteListener<EndTransitionAction, EffectAPI>
  ): UntouchedRouteEffects<EffectAPI, Handlers>;

  /**
   * Adds a route listener that's called when a route transition
   * for the given route interrupts.
   *
   * @example
   *
   * const routes = createRoutes()
   *   .set("root", "/");
   *
   * createRouteEffects()
   *   .onInterrupt(routes.Root.id, () => {
   *     console.log("The root route transition was interrupted.");
   *   });
   *
   * @public
   */
  onInterrupt<RID extends RouteID>(
    routeInput: RouteInput<RID>,
    effect: RouteListener<EndTransitionAction, EffectAPI>
  ): UntouchedRouteEffects<EffectAPI, Handlers>;

  /**
   * Adds a route listener that's called when a route transition
   * for any route interrupts.
   *
   * @example
   *
   * const routes = createRoutes()
   *   .set("root", "/");
   *
   * createRouteEffects()
   *   .onInterruptAll(() => {
   *     console.log("A route transition was interrupted.");
   *   });
   *
   * @public
   */
  onInterruptAll(
    effect: RouteListener<EndTransitionAction, EffectAPI>
  ): UntouchedRouteEffects<EffectAPI, Handlers>;

  /**
   * Adds a route listener that's called when a route transition
   * for the given route starts.
   *
   * @example
   *
   * const routes = createRoutes()
   *   .set("root", "/");
   *
   * createRouteEffects()
   *   .onStart(routes.Root.id, () => {
   *     console.log("The root route transition started.");
   *   });
   *
   * @public
   */
  onStart<RID extends RouteID>(
    routeInput: RouteInput<RID>,
    effect: RouteListener<StartTransitionAction, EffectAPI>
  ): UntouchedRouteEffects<EffectAPI, Handlers>;

  /**
   * Adds a route listener that's called when a route transition
   * for any route starts.
   *
   * @example
   *
   * const routes = createRoutes()
   *   .set("root", "/");
   *
   * createRouteEffects()
   *   .onStartAll(() => {
   *     console.log("A route transition started.");
   *   });
   *
   * @public
   */
  onStartAll(
    effect: RouteListener<StartTransitionAction, EffectAPI>
  ): UntouchedRouteEffects<EffectAPI, Handlers>;

  [SYMBOL_ROUTE_EFFECTS_META]: RouteEffectsMeta<EffectAPI>;
}

type HandlerMetaRID<H> = H extends HandlerMeta<
  RouteEffectHookName,
  infer RID,
  AnyPayloadAction
>
  ? RID
  : never;
type HandlerMetaHookName<H> = H extends HandlerMeta<
  infer HN,
  RouteID,
  AnyPayloadAction
>
  ? HN
  : never;
type HandlerMetaAction<H> = H extends HandlerMeta<
  RouteEffectHookName,
  RouteID,
  infer Action
>
  ? Action
  : never;

type RouteMiddlewareReference<
  RID extends RouteID,
  EffectAPI extends DefaultEffectAPI,
  Action extends AnyPayloadAction
> = RouteMiddleware<Action, EffectAPI> & {
  readonly id: RID;
  readonly type: RouteEffectHookName;
};

type RouteListenerReference<
  RID extends RouteID,
  EffectAPI extends DefaultEffectAPI,
  Action extends AnyPayloadAction
> = RouteListener<Action, EffectAPI> & {
  readonly id: RID;
  readonly eventName: RouteEffectHookName;
};

type RouteEffectsHandlers<
  EffectAPI extends DefaultEffectAPI,
  Handlers extends HandlerMeta<
    RouteEffectHookName,
    RouteID,
    AnyPayloadAction
  > = HandlerMeta<RouteEffectHookName, never, PayloadAction<never>>
> = {
  readonly [K in Handlers as HandlerName<
    HandlerMetaHookName<K>,
    HandlerMetaRID<K>
  >]: RouteMiddlewareReference<
    HandlerMetaRID<K>,
    EffectAPI,
    HandlerMetaAction<K>
  >;
};

type ReadonlyRouteEffectsMethods<EffectAPI extends DefaultEffectAPI> = {
  /**
   * Retrieves a map of handlers to invoke effects matching
   * the given route ID.
   *
   * @privateRemarks To ensure no parameters are accidentally left out,
   * a union type with `undefined` is used instead of making the parameters optional.
   *
   * @internal
   */
  _getHandlersByRouteID(
    id: RouteID | undefined
  ): Promise<RouteEffectHandlers<EffectAPI>>;
};

/** @internal */
export type UntouchedRouteEffects<
  EffectAPI extends DefaultEffectAPI,
  Handlers extends HandlerMeta<
    RouteEffectHookName,
    RouteID,
    AnyPayloadAction
  > = HandlerMeta<RouteEffectHookName, never, PayloadAction<never>>
> = RouteEffectsMethods<EffectAPI, Handlers> &
  ReadonlyRouteEffectsMethods<EffectAPI> & {
    handlers: RouteEffectsHandlers<EffectAPI, Handlers>;
  };

/**
 * @see {createRouteEffects}
 *
 * @public
 */
export type RouteEffects<EffectAPI extends DefaultEffectAPI> = {
  [SYMBOL_ROUTE_EFFECTS_META]: RouteEffectsMeta<EffectAPI>;
} & ReadonlyRouteEffectsMethods<EffectAPI>;

/** @internal */
type RouteInput<RID extends RouteID> = RID | { id: RID };

/** @internal */
function resolveRouteId<RID extends RouteID>(routeInput: RouteInput<RID>) {
  return typeof routeInput == "string" ? routeInput : routeInput.id;
}

function getHandlerName<RID extends RouteID, H extends RouteEffectHookName>(
  id: RID,
  hookName: H
): HandlerName<H, RID> {
  return `${capitalize(id)}${capitalize(hookName)}`;
}

/**
 * Creates an object with a fluent interface for adding effects to route transitions.
 * A Route Effects object is essentially a list of effects assigned to one or all routes.
 *
 * Route Effects objects methods can be chained, and return a new object when called.
 *
 * Two types of effects can be added, middleware and listeners.
 *
 * Middleware effects are called asynchronously in a series within a route transition.
 * A middleware effect must resolve before the next middleware effect is invoked.
 *
 * Listener effects are called synchronously in a series outside of the route transition.
 * A listener effect may be asynchronous, but they're not awaited.
 *
 * @example
 *
 * const initialRoutes = createRoutes()
 *   .set("root", "/")
 *   .setWildcard("profiles", "/profiles/*")
 *   .set("legacyRoot", "/app");
 *
 * createRouteEffects()
 *   // Delay all route transitions by 500ms
 *   .appendAll(delayTransition(500))
 *   // Forward "legacyRoot" to the "root".
 *   // This effect will called before the transition delay.
 *   .prepend("legacyRoot", forwardTo("/"))
 *   // Lazy-load routes when a location matches the Profiles wildcard route
 *   .append(
 *     initialRoutes.Profiles,
 *     addRoutes(() => import("./profileRoutes.js"))
 *   );
 *
 * @public
 */
export function createRouteEffects<
  EffectAPI extends DefaultEffectAPI,
  Handlers extends HandlerMeta<
    RouteEffectHookName,
    RouteID,
    RouteEffectAction
  > = HandlerMeta<RouteEffectHookName, never, RouteEffectAction>
>() {
  function getRouteMiddlewareHandler(
    routeHookName: RouteEffectHookName,
    routeID?: RouteID
  ) {
    const handlerEffects: RouteMiddlewareReference<string, EffectAPI, any>[] =
      [];
    const currentEffects = [
      ...routeEffects[SYMBOL_ROUTE_EFFECTS_META].middleware,
    ];

    for (const { middlewareReference } of currentEffects) {
      const { id, type: hookName } = middlewareReference;

      if (
        hookName === routeHookName &&
        (id === ALL_ROUTES_KEY || id === routeID)
      ) {
        handlerEffects.push(middlewareReference);
      }
    }

    return chainMiddleware(...handlerEffects);
  }

  function getRouteListenerHandler(
    routeHookName: RouteEffectHookName,
    routeID?: RouteID
  ) {
    const handlerListeners: RouteListenerReference<string, EffectAPI, any>[] =
      [];
    const currentListeners = [
      ...routeEffects[SYMBOL_ROUTE_EFFECTS_META].listeners,
    ];

    for (const { listenerReference } of currentListeners) {
      const { id, eventName: hookName } = listenerReference;

      if (
        hookName === routeHookName &&
        (id === ALL_ROUTES_KEY || id === routeID)
      ) {
        handlerListeners.push(listenerReference);
      }
    }

    return combineListeners(...handlerListeners);
  }

  const addEffectModes = {
    APPEND: "append",
    PREPEND: "prepend",
  } as const;

  type AddEffectMode = typeof addEffectModes[keyof typeof addEffectModes];

  function _clone(
    mutateRouteEffectMeta?: (
      nextRouteEffectsMeta: RouteEffectsMeta<EffectAPI>
    ) => void
  ): UntouchedRouteEffects<EffectAPI, Handlers> {
    function createNextRouteEffectsMeta(): RouteEffectsMeta<EffectAPI> {
      const routeEffectMeta = routeEffects[SYMBOL_ROUTE_EFFECTS_META];
      const nextRouteEffectsMeta: RouteEffectsMeta<EffectAPI> = {
        middleware: new Set(routeEffectMeta.middleware),
        listeners: new Set(routeEffectMeta.listeners),
      };

      mutateRouteEffectMeta?.(nextRouteEffectsMeta);

      return nextRouteEffectsMeta;
    }

    function createHandlers(nextRouteEffectsMeta: RouteEffectsMeta<EffectAPI>) {
      return Object.fromEntries(
        [...nextRouteEffectsMeta.middleware].map(
          ({ referenceKey, middlewareReference }) => {
            const { id, type } = middlewareReference;
            const handler = getRouteMiddlewareHandler(type, id);

            return [referenceKey, Object.assign(handler, { id, type })];
          }
        )
      ) as RouteEffectsHandlers<EffectAPI, Handlers>;
    }

    const nextRouteEffects = createRouteEffects<EffectAPI, Handlers>();
    const nextRouteEffectsMeta = createNextRouteEffectsMeta();

    nextRouteEffects[SYMBOL_ROUTE_EFFECTS_META] = nextRouteEffectsMeta;
    nextRouteEffects.handlers = createHandlers(nextRouteEffectsMeta);

    return nextRouteEffects;
  }

  const routeEffects = {
    _clone,
    append(routeInput, effect) {
      return addMiddleware(
        addEffectModes.APPEND,
        routeInput,
        hookNames.MIDDLEWARE,
        effect
      );
    },
    appendAll(effect) {
      return addMiddleware(
        addEffectModes.APPEND,
        ALL_ROUTES_KEY,
        hookNames.MIDDLEWARE,
        effect
      );
    },
    prepend(routeInput, effect) {
      return addMiddleware(
        addEffectModes.PREPEND,
        routeInput,
        hookNames.MIDDLEWARE,
        effect
      );
    },
    prependAll(effect) {
      return addMiddleware(
        addEffectModes.PREPEND,
        ALL_ROUTES_KEY,
        hookNames.MIDDLEWARE,
        effect
      );
    },
    onComplete(routeInput, listener) {
      return addListener(routeInput, hookNames.COMPLETE, listener);
    },
    onCompleteAll(listener) {
      return addListener(ALL_ROUTES_KEY, hookNames.COMPLETE, listener);
    },
    onEnd(routeInput, listener) {
      return addListener(routeInput, hookNames.END, listener);
    },
    onEndAll(listener) {
      return addListener(ALL_ROUTES_KEY, hookNames.END, listener);
    },
    onExit(routeInput, listener) {
      return addListener(routeInput, hookNames.EXIT, listener);
    },
    onExitAll(listener) {
      return addListener(ALL_ROUTES_KEY, hookNames.EXIT, listener);
    },
    onFailure(routeInput, listener) {
      return addListener(routeInput, hookNames.FAILURE, listener);
    },
    onFailureAll(listener) {
      return addListener(ALL_ROUTES_KEY, hookNames.FAILURE, listener);
    },
    onInterrupt(routeInput, listener) {
      return addListener(routeInput, hookNames.INTERRUPT, listener);
    },
    onInterruptAll(listener) {
      return addListener(ALL_ROUTES_KEY, hookNames.INTERRUPT, listener);
    },
    onStart(routeInput, listener) {
      return addListener(routeInput, hookNames.START, listener);
    },
    onStartAll(listener) {
      return addListener(ALL_ROUTES_KEY, hookNames.START, listener);
    },
    async _getHandlersByRouteID(id) {
      return {
        middleware: getRouteMiddlewareHandler(hookNames.MIDDLEWARE, id),
        onComplete: getRouteListenerHandler(hookNames.COMPLETE, id),
        onEnd: getRouteListenerHandler(hookNames.END, id),
        onExit: getRouteListenerHandler(hookNames.EXIT, id),
        onFailure: getRouteListenerHandler(hookNames.FAILURE, id),
        onInterrupt: getRouteListenerHandler(hookNames.INTERRUPT, id),
        onStart: getRouteListenerHandler(hookNames.START, id),
      };
    },
    [SYMBOL_ROUTE_EFFECTS_META]: {
      middleware: new Set(),
      listeners: new Set(),
    },
  } as UntouchedRouteEffects<EffectAPI, Handlers>;

  function wrapFunction<T extends (...args: any[]) => any>(fn: T) {
    return function wrappedFn(...args: Parameters<T>) {
      return fn(...args);
    };
  }

  function addMiddleware<RID extends RouteID, Action extends RouteEffectAction>(
    mode: AddEffectMode,
    routeInput: RouteInput<RID>,
    hookName: RouteEffectHookName,
    effect: RouteMiddleware<Action, EffectAPI>
  ) {
    const id = resolveRouteId(routeInput);
    const referenceKey = getHandlerName(id, hookName);
    const nextRouteEffects = routeEffects._clone((nextRouteEffectsMeta) => {
      const effectToAdd = {
        referenceKey,
        middlewareReference: Object.assign(wrapFunction(effect), {
          id,
          type: hookName,
        }),
      };
      const middleware = [...nextRouteEffectsMeta.middleware];

      if (mode === addEffectModes.APPEND) {
        middleware.push(effectToAdd);
      } else if (mode === addEffectModes.PREPEND) {
        middleware.unshift(effectToAdd);
      }

      nextRouteEffectsMeta.middleware = new Set(middleware);
    });

    return nextRouteEffects;
  }

  function addListener<RID extends RouteID, Action extends RouteEffectAction>(
    routeInput: RouteInput<RID>,
    hookName: RouteEffectHookName,
    listener: RouteListener<Action, EffectAPI>
  ) {
    const id = resolveRouteId(routeInput);
    const nextRouteEffects = routeEffects._clone((nextRouteEffectsMeta) => {
      nextRouteEffectsMeta.listeners.add({
        listenerReference: Object.assign(wrapFunction(listener), {
          id,
          eventName: hookName,
        }),
      });
    });

    return nextRouteEffects;
  }

  return routeEffects as UntouchedRouteEffects<EffectAPI, Handlers>;
}

/**
 * Determines whether the given value is a Route Effects object
 *
 * @public
 */
export function isRouteEffects<R extends RouteEffects<DefaultEffectAPI>>(
  value: unknown
): value is R {
  return (
    typeof value == "object" &&
    value !== null &&
    Object.hasOwn(value, SYMBOL_ROUTE_EFFECTS_META)
  );
}
