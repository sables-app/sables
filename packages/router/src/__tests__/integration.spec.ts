import * as Redux from "redux";
import { interval, lastValueFrom, map, take } from "rxjs";
import * as vitest from "vitest";
import { describe, expect, it, test, vi } from "vitest";

import {
  advanceLocation,
  ensureLocation,
  locationChange,
  retreatLocation,
} from "../actions.js";
import { endRouteTransitionReasons } from "../constants.js";
import { createRouteEffects } from "../RouteEffects.js";
import { createRoutes } from "../Routes.js";
import { selectCurrentRoute } from "../selectors.js";
import { buildHref } from "../utils.js";
import {
  addInvokedImportersToEffectAPI,
  initializeTestRouter,
  mockImporter,
  waitForEnsuredRouteTransition,
  waitForRouteTransition,
} from "./utils.js";

vi.stubEnv("SSR", "true");

describe("integration", () => {
  test("setRoutes", async () => {
    let invokedImporters: Set<string> | undefined = undefined;
    const aboutDogsRouteEffect = vi.fn();
    const aboutRouteEffectsAddListener = vi.fn();
    const aboutRouteEffects = createRouteEffects()
      .append("aboutDogs", aboutDogsRouteEffect)
      .onAdd(aboutRouteEffectsAddListener);

    const aboutRouteEffectsImporter = mockImporter(
      vitest,
      "aboutRouteEffectsImporter",
      aboutRouteEffects,
    );
    const aboutRoutes = createRoutes()
      .set("aboutUs", "/app/about/company")
      .set("aboutDogs", "/app/about/dogs")
      .setEffects(aboutRouteEffectsImporter);
    const aboutRoutesImporter = mockImporter(
      vitest,
      "aboutRoutesImporter",
      aboutRoutes,
    );
    const interruptListener = vi.fn();
    const initialRouteEffects =
      createRouteEffects().onInterruptAll(interruptListener);
    const initialRoutes = createRoutes()
      .set("appRoot", "/app")
      .set("dogs", "/app/dogs")
      .setRoutes("/app/about/*", aboutRoutesImporter)
      .setEffects(() => initialRouteEffects);
    const { actions$, assertCurrentRouteID, store } = initializeTestRouter(
      vitest,
      {
        initialLocation: "/app",
        initialRoutes,
        effectAPI(defaults) {
          invokedImporters = addInvokedImportersToEffectAPI(defaults);
          return defaults;
        },
      },
    );

    expect(selectCurrentRoute(store.getState())).toBeNull();
    {
      // Wait for the initial route transition to complete
      await waitForRouteTransition();
      // Wait for `initialRouteEffects` to be resolved
      await Promise.resolve();
      // Wait for `initialRouteEffects.onAdd` to be resolved
      await Promise.resolve();
    }
    assertCurrentRouteID(initialRoutes.AppRoot.id);

    const aboutDogsHref = buildHref(aboutRoutes.AboutDogs);

    expect(aboutDogsHref).toBe("/app/about/dogs");
    {
      store.dispatch(ensureLocation(aboutDogsHref));
    }
    expect(aboutRoutesImporter).not.toHaveBeenCalled();
    {
      // Wait for the about routes to be imported
      await waitForEnsuredRouteTransition();
    }
    expect(aboutRoutesImporter).toHaveBeenCalledOnce();
    expect(invokedImporters).toHaveProperty("size", 1);
    {
      // Wait for history replacement transition to start
      await waitForRouteTransition();
    }
    expect(aboutRouteEffectsImporter).toHaveBeenCalledOnce();
    {
      // Wait for the importer to resolve
      await Promise.resolve();
    }
    expect(invokedImporters).toHaveProperty("size", 2);
    expect(invokedImporters).toMatchSnapshot();
    expect(aboutDogsRouteEffect).not.toHaveBeenCalled();
    expect(aboutRouteEffectsAddListener).not.toHaveBeenCalled();
    {
      // Wait for route effects add handler to resolve
      await Promise.resolve();
      // Wait for route effects add handler to be invoked
      await Promise.resolve();
    }
    expect(aboutRouteEffectsAddListener).toHaveBeenCalledOnce();
    {
      // Wait for history replacement transition to end
      await waitForRouteTransition();
    }
    expect(interruptListener).toHaveBeenCalledOnce();
    expect(interruptListener).toHaveBeenCalledWith(
      expect.objectContaining({
        type: "sables/transitionRoute/end",
        payload: {
          locationChange: expect.any(Object),
          reason: endRouteTransitionReasons.INTERRUPTED,
          transitionID: expect.any(String),
        },
      }),
      expect.objectContaining({
        actions$,
      }),
    );
    expect(aboutDogsRouteEffect).toHaveBeenCalledOnce();
    {
      // Wait two microtasks for `aboutDogsRouteEffect` to resolve
      await Promise.resolve().then(() => Promise.resolve());
    }
    assertCurrentRouteID(aboutRoutes.AboutDogs.id);
  });

  test("retreatLocation and advanceLocation", async () => {
    const initialRoutes = createRoutes()
      .set("appRoot", "/app")
      .set("dogs", "/app/dogs")
      .set("cats", "/app/cats");
    const { assertCurrentRouteID, store } = initializeTestRouter(vitest, {
      initialLocation: "/app",
      initialRoutes,
    });

    expect(selectCurrentRoute(store.getState())).toBeNull();
    {
      await waitForRouteTransition();
    }
    assertCurrentRouteID(initialRoutes.AppRoot.id);
    {
      store.dispatch(ensureLocation(initialRoutes.Cats));
      await waitForEnsuredRouteTransition();
    }
    assertCurrentRouteID(initialRoutes.Cats.id);
    {
      store.dispatch(ensureLocation(initialRoutes.Dogs));
      await waitForEnsuredRouteTransition();
    }
    assertCurrentRouteID(initialRoutes.Dogs.id);
    {
      store.dispatch(retreatLocation());
      await waitForRouteTransition();
    }
    assertCurrentRouteID(initialRoutes.Cats.id);
    {
      store.dispatch(advanceLocation());
      await waitForRouteTransition();
    }
    assertCurrentRouteID(initialRoutes.Dogs.id);
  });

  describe("when multiple equivalent `ensureLocation` actions are dispatched", () => {
    function prepareTest() {
      const initialRoutes = createRoutes()
        .set("appRoot", "/app")
        .set("dogs", "/app/dogs")
        .set("cats", "/app/cats");
      const { actions$, assertCurrentRouteID, store } = initializeTestRouter(
        vitest,
        {
          initialLocation: "/app",
          initialRoutes,
        },
      );

      const dispatchedActions: Redux.AnyAction[] = [];

      actions$.subscribe((action) => dispatchedActions.push(action));

      return { assertCurrentRouteID, dispatchedActions, initialRoutes, store };
    }

    function testOneLocationPush(
      testTransition: (
        testBag: ReturnType<typeof prepareTest>,
      ) => Promise<void>,
    ) {
      it("dispatches only one location push action", async () => {
        const testBag = prepareTest();
        const {
          assertCurrentRouteID,
          dispatchedActions,
          initialRoutes,
          store,
        } = testBag;

        expect(selectCurrentRoute(store.getState())).toBeNull();
        {
          await waitForRouteTransition();
        }
        assertCurrentRouteID(initialRoutes.AppRoot.id);
        {
          await testTransition(testBag);
        }
        assertCurrentRouteID(initialRoutes.Cats.id);

        expect(dispatchedActions.filter(ensureLocation.match)).toHaveLength(4);
        expect(
          dispatchedActions.filter(locationChange.isPushAction),
        ).toHaveLength(1);
      });
    }

    describe("synchronously", () => {
      testOneLocationPush(async ({ store, initialRoutes }) => {
        store.dispatch(ensureLocation(initialRoutes.Cats));
        store.dispatch(ensureLocation(initialRoutes.Cats));
        store.dispatch(ensureLocation(initialRoutes.Cats));
        store.dispatch(ensureLocation(initialRoutes.Cats));
        await waitForEnsuredRouteTransition();
      });
    });

    describe("asynchronously", () => {
      testOneLocationPush(async ({ store, initialRoutes }) => {
        const dispatches = lastValueFrom(
          // Using 50ms to simulate an end-user clicking a link really fast
          interval(50).pipe(
            take(4),
            map(() => {
              store.dispatch(ensureLocation(initialRoutes.Cats));
            }),
          ),
        );

        await Promise.all([dispatches, waitForEnsuredRouteTransition()]);
      });
    });
  });
});
