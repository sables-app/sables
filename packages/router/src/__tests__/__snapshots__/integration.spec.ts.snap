// Vitest Snapshot v1

exports[`integration > setRoutes 1`] = `
Set {
  "aboutRoutesImporter",
  "aboutRouteEffectsImporter",
}
`;
